### PR TITLE
chore: rewrite schemas package without typing internals

### PR DESCRIPTION
**Pull request by rb-determined-ai**

### CODE REVIEW

File: harness/determined/common/schemas/_schema_base.py [hunk 1]
```diff
@@ -1,13 +1,45 @@
...
-import typing
-from typing import Any, List, Mapping, Optional, Sequence, Type, TypeVar, cast
+from typing import Any, Dict, List, Mapping, Optional, Sequence, Type, TypeVar
 
 from determined.common import schemas
 from determined.common.schemas import expconf
 
 PRIMITIVE_JSON_TYPES = (numbers.Number, str, bool, type(None))
 
+# `typing` has some awkward APIs for type inspection.  In general, the only thing we can rely on
```
Comment (ioga) [inline harness/determined/common/schemas/_schema_base.py:10]: when can we make it simpler? after retiring 3.6, 3.7, ...?

**Review by ioga — APPROVED** [review]
where did duck typing

File: harness/determined/common/schemas/_schema_base.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/determined/common/schemas/_schema_base.py:10]: my guess is "when we decide to remove _instance_from_annotation and come up with a whole different solution".

In 3.6 you could check if a given annotation was some sort of Union, but I can't find any API for that in 3.7+, so all I could come up with was checking equality against a list of whitelisted Union types.

**Comment by rb-determined-ai** [comment]
> where did duck typing [go?]

If we subclassed all of the primitive types (`int`, `str`, `bool`, etc) then we could do an easy duckt-yping solution.  I was never willing to consider that though, even in the Go code.

**Pull request merged**